### PR TITLE
Fix _alias/<alias> returning non-matching data streams

### LINKED ISSUE #96589: `GET _alias/<alias>` returns aliases for non-matching data streams
### Elasticsearch Version

8.9.0-SNAPSHOT

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

MacOS 13.2.1 (arm64)

### Problem Description

In 8.9.0-SNAPSHOT if a data stream `my-data-stream` is created and has the alias `my-alias` it is returned when running `GET _alias/foobarbaz*`. 

My expectation is that only data streams starting with “foobarbaz” should be returned. 

### Steps to Reproduce

```
--
# Create index template
PUT _index_template/my-index-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": { },
  "priority": 500
}

--
# Create data stream
PUT _data_stream/my-data-stream

--
# Create alias for data stream
POST _aliases
{
  "actions": [
    {
      "add": {
        "index": "my-data-stream",
        "alias": "my-alias"
      }
    }
  ]
}

--
# List aliases
GET _alias/foobarbaz*
```

This returns:

```
{
  "my-data-stream": {
    "aliases": {
      "my-alias": {}
    }
  }
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by mattc58**
This fixes a bug where `GET _alias/<alias>` would return non-matching data streams. This happens when security is enabled and when the index pattern is changed to `["*',"-*"]` and the data streams don't properly process that.

Fixes #96589.

### CODE REVIEW

**Comment by mattc58** [comment]
The new REST test can be run with: `./gradlew ':modules:data-streams:yamlRestTest' --tests "org.elasticsearch.datastreams.DataStreamsClientYamlTestSuiteIT.test {p0=data_stream/140_data_stream_aliases/Test get alias with non-matching data streams}" --info`

It will fail with the current code with:
```
      - is_false: ds-first.aliases.my-alias

    [2024-01-10T03:34:49,505][INFO ][o.e.d.DataStreamsClientYamlTestSuiteIT] [test] Stash dump on test failure [{
      "stash" : {
        "body" : {
          "ds-first" : {
            "aliases" : {
              "my-alias" : { }
            }
          }
        }
      }
    }]

[snip]

        Caused by:
        java.lang.AssertionError: field [ds-first.aliases.my-alias] has a true value but it shouldn't
        Expected: ("" or a string equal to "false" ignoring case or "0")
             but: was "{}"
            at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
            at org.junit.Assert.assertThat(Assert.java:956)
            at org.elasticsearch.test.rest.yaml.section.IsFalseAssertion.doAssert(IsFalseAssertion.java:48)
            at org.elasticsearch.test.rest.yaml.section.Assertion.execute(Assertion.java:65)
            at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:558)
            ... 38 more
```

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataTests.java [hunk 1]
```diff
@@ -155,6 +155,55 @@ public void testFindAliases() {
...
+                request.aliases(),
+                new String[] { "index", "d1", "d2" }
+            );
+            assertEquals(1, aliases.size());
+            List<DataStreamAlias> found = aliases.get("d1");
+            assertThat(found, transformedItemsMatch(DataStreamAlias::getAlias, contains("alias1")));
+        }
+
+        {
+            GetAliasesRequest request = new GetAliasesRequest();
```
Comment (masseyke) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetadataTests.java:198]: Should we have a test in here that covers the exclusions case (like `testFindAliasWithExclusion` but for findDataStreamAliases? I don't know that we even claim to support it, but we get it virtually for free with what you've done.

File: server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java [hunk 2]
```diff
@@ -147,20 +146,17 @@ static Map<String, List<DataStreamAlias>> postProcess(
...
-            List<DataStreamAlias> aliases = state.metadata()
-                .dataStreamAliases()
-                .values()
-                .stream()
-                .filter(alias -> alias.getDataStreams().contains(requestedDataStream))
-                .filter(alias -> noAliasesSpecified || Regex.simpleMatch(request.aliases(), alias.getName()))
-                .toList();
-            if (aliases.isEmpty() == false) {
-                result.put(requestedDataStream, aliases);
+            if (dsAliases.containsKey(requestedDataStream)) {
```
Comment (masseyke) [inline server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:155]: I think that this check is redundant now, right? In Metadata, the setter only puts things in the map that's returned if they are the things you passed in as the second argument to `findDataStreamAliases`.

Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:155]: That is, can you just return the result of `findDataStreamAliases` from this method?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 3]
```diff
@@ -799,11 +809,63 @@ public Map<String, List<AliasMetadata>> findAllAliases(final String[] concreteIn
...
+     * that point to the specified data streams (directly or matching data streams via wildcards).
+     *
+     * @param aliases The aliases to look for. Might contain include or exclude wildcards.
+     * @param dataStreams The data streams that the aliases must point to in order to be returned
+     * @return A map of data stream name to the list of DataStreamAlias objects that match. If a data stream does not have matching
+     * aliases then the result will <b>not</b> include the data stream's key.
+     */
+    public Map<String, List<DataStreamAlias>> findDataStreamAliases(final String[] aliases, final String[] dataStreams) {
+        ImmutableOpenMap.Builder<String, List<DataStreamAlias>> mapBuilder = ImmutableOpenMap.builder();
+        Map<String, List<DataStreamAlias>> dataStreamAliases = dataStreamAliasesByDataStream();
```
Comment (masseyke) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:838]: I'm wondering if at some point we'll need to start computing this outside of this method if users have a _lot_ of data stream aliases. But doing it once per request doesn't seem all that bad.

**Review by masseyke — APPROVED** [review]
LGTM. Seems like a big improvement over what we had.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/140_data_stream_aliases.yml [hunk 4]
```diff
@@ -260,3 +260,50 @@
   - match: {test2.aliases: {}}
   - match: {test3.aliases: {}}
 
+---
+"Test get alias with non-matching data streams":
+  - skip:
+      version: " - 8.12.1"
```
Comment (masseyke) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/140_data_stream_aliases.yml:266]: This is currently pointed at main (8.13.0) and the PR isn't labeled with 8.12.1. We'll need to make sure to remember backport it, or change this version if this doesn't get backported in time for 8.12.1

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 3 above)
Comment (mattc58) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:838]: I wonder if doing the analysis and guidelines suggested in https://github.com/elastic/elasticsearch/issues/104456 would help guide us here. Maybe we could store aliases differently in cluster state.

**Comment by mattc58** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

**Pull request merged**